File: bindings/react-native/src/facets/dataframe.ts
import UldrenLoom from '../NativeUldrenLoom';
import { authArgs, keyArgs } from '../internal';
import type { LoomAuth, LoomKey } from '../internal';

/** Create dataframe frame `name` from canonical DataframePlan CBOR. */
export function dataframeCreate(
  loomPath: string,
  workspace: string,
  name: string,
  plan: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.dataframeCreate(
    loomPath,
    workspace,
    name,
    Array.from(plan),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** Execute dataframe frame `name` and return canonical CBOR `[columns, rows]`. */
export async function dataframeCollectCbor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.dataframeCollect(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** Execute dataframe frame `name` and return at most `rows` rows as canonical CBOR `[columns, rows]`. */
export async function dataframePreviewCbor(
  loomPath: string,
  workspace: string,
  name: string,
  rows: number,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.dataframePreview(loomPath, workspace, name, rows, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** Materialize dataframe frame `name`; returns a CAS digest when the materialization target emits one. */
export function dataframeMaterialize(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<string | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.dataframeMaterialize(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase);
}

/** Canonical dataframe plan digest as `algo:hex`. */
export function dataframePlanDigest(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<string> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.dataframePlanDigest(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase);
}

/** Source digests pinned in the dataframe plan as canonical CBOR array of `algo:hex` strings. */
export async function dataframeSourceDigestsCbor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.dataframeSourceDigests(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase)
  );
}

File: bindings/android/src/commonMain/kotlin/ai/uldren/loom/facets/dataframe.kt
package ai.uldren.loom

/** Create dataframe frame [name] from canonical DataframePlan CBOR [plan]. */
expect fun Loom.dataframeCreate(
        path: String,
        workspace: String,
        name: String,
        plan: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )

/** Execute dataframe frame [name] and return canonical CBOR `[columns, rows]`. */
expect fun Loom.dataframeCollectCbor(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray

/** Execute dataframe frame [name] and return at most [rows] rows as canonical CBOR `[columns, rows]`. */
expect fun Loom.dataframePreviewCbor(
        path: String,
        workspace: String,
        name: String,
        rows: Long,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray

/** Materialize dataframe frame [name]; returns a CAS digest when the materialization target emits one. */
expect fun Loom.dataframeMaterialize(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): String?

/** Canonical dataframe plan digest as `algo:hex`. */
expect fun Loom.dataframePlanDigest(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): String

/** Source digests pinned in the dataframe plan as canonical CBOR array of `algo:hex` strings. */
expect fun Loom.dataframeSourceDigestsCbor(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray
